readium/adapters/pspdfkit/pspdfkit-navigator/src/main/java/org/readium/adapters/pspdfkit/navigator/PsPdfKitDocumentFragment.kt

```diff
@@ -31,35 +35,92 @@ import com.pspdfkit.listeners.OnPreparePopupToolbarListener
 import org.readium.r2.shared.publication.Publication
 import org.readium.r2.shared.publication.services.isProtected
+import org.readium.r2.shared.resource.Resource
+import org.readium.r2.shared.resource.ResourceTry
+import org.readium.r2.shared.util.Url
+import org.readium.r2.shared.util.pdf.cachedIn
 
 @ExperimentalReadiumApi
-internal class PsPdfKitDocumentFragment(
+public class PsPdfKitDocumentFragment(
     private val publication: Publication,
-    private val document: PsPdfKitDocument,
-    private val initialPageIndex: Int,
-    settings: PsPdfKitSettings,
-    private val listener: Listener?
-) : PdfDocumentFragment<PsPdfKitSettings>() {
+    private val href: Url,
+    initialPageIndex: Int,
+    initialSettings: PsPdfKitSettings,
+    private val listener: Listener?,
+    private val inputListener: InputListener
+) : PdfDocumentFragment<PsPdfKitDocumentFragment.Listener, PsPdfKitSettings>() {
+
+    public interface Listener : PdfDocumentFragment.Listener {
+        /**
+         * Called when a PDF resource failed to be loaded, for example because of an
+         * [OutOfMemoryError].
+         */
+        public fun onResourceLoadFailed(href: Url, error: Resource.Exception) {}
+
+        /** Called when configuring a new PDF fragment. */
+        public fun onConfigurePdfView(builder: PdfConfiguration.Builder): PdfConfiguration.Builder = builder
+    }
+
+    private companion object {
+        private const val pdfFragmentTag = "com.pspdfkit.ui.PdfFragment"
+    }
+
+    private var pdfFragment: PdfFragment? = null
+    private val psPdfKitListener = PsPdfKitListener()
+
+    private class DocumentViewModel(
+        document: suspend () -> ResourceTry<PsPdfKitDocument>
+    ) : ViewModel() {
 
-    override var settings: PsPdfKitSettings = settings
-        set(value) {
-            if (field == value) return
+        val document: Deferred<ResourceTry<PsPdfKitDocument>> =
+            viewModelScope.async { document() }
+    }
 
-            field = value
-            reloadDocumentAtPage()
+    private val viewModel: DocumentViewModel by viewModels {
+        createViewModelFactory {
+            DocumentViewModel(
+                document = {
+                    PsPdfKitDocumentFactory(requireContext())
+                        .cachedIn(publication)
+                        .open(publication.get(href), null)
+                }
+            )
         }
+    }
 
-    private lateinit var pdfFragment: PdfFragment
-    private val psPdfKitListener = PsPdfKitListener()
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        // We don't support fragment restoration for the PdfFragment, as we want to recreate a fresh
+        // instance in [reset]. To prevent restoring (and crashing) the PdfFragment without a
+        // document source, we remove it from the fragment manager.
+        (childFragmentManager.findFragmentByTag(pdfFragmentTag) as? PdfFragment)
```

Comment:
Thankfully, PSPDFKit's `PdfFragment` will crash when creating its view and not when creating the fragment, so we have time to remove it from the fragment manager.